### PR TITLE
Add support for disabling default firewalls and opening up ports that can be accessed remotely.

### PR DESCRIPTION
**Pull request by kwsimons**
.NET port of landaire's hard work: https://github.com/exploits-forsale/solstice/blob/main/crates/solstice_daemon/src/firewall.rs

---

Tested on an Xbox One X version 10.0.25398.4909:
1. Copy .NET to `D:\XBOX\dotnet\`
2. Copy this DLL (`DurangoInteropDotnet.dll`) to `D:\XBOX\payloads\`
3. Copy `msbuild_tasks/disable_firewall.xml` to `D:\XBOX\payloads\disable_firewall.xml`
4. Execute the following commands:
```
	set DOTNET_CLI_TELEMETRY_OPTOUT=1
	set DOTNET_EnableWriteXorExecute=0
	D:\XBOX\dotnet\dotnet.exe msbuild "D:\XBOX\payloads\disable_firewall.xml"
```
5. Expected output:
```
MSBuild version 17.10.4+10fbfbf2e for .NET
Disabling firewall and opening up SSH/Debug ports ... done.

### CODE REVIEW

**Review by kwsimons — COMMENTED** [review]
Note: The FWPM Provider persists across reboots, but the firewall rules do not.

File: src/FirewallManager.cs [hunk 1]
```diff
@@ -0,0 +1,1096 @@
...
+        private static readonly string FWPM_DISPLAY_NAME = "XOR";
+        private static readonly string FWPM_DISPLAY_DESCRIPTION = "XOR FWPM Provider";
+        private static readonly Guid FWPM_PROVIDER_GUID = new Guid(0xabad1dea, 0x4141, 0x4141, 0x0, 0x0, 0x0c, 0x0f, 0xfe, 0xe0, 0x00, 0x00);
+        private static readonly Guid CLSID_HNETCFGFWPOLICY2 = new Guid("e2b3c97f-6ae1-41ac-817a-f6f92166d7dd");
+        private static readonly Guid CLSID_INETFWPOLICY2 = new Guid("98325047-C671-4174-8D81-DEFCD3F03186");
+
+        public static void DisableFirewalls() {
+
+            uint result = NativeBridge.CoInitializeEx(IntPtr.Zero, NativeBridge.COINIT.COINIT_MULTITHREADED);
+            if (result != 0 && result != 1) { // 0 = S_OK, 1 = S_FALSE
```
Comment (tuxuser) [inline src/FirewallManager.cs:40]: This looks a bit odd?

**Review by tuxuser — CHANGES_REQUESTED** [review]
Please check the one commented line.

Update: thanks for your contribution 🙂

File: src/FirewallManager.cs (hunk 1 above)
Comment (kwsimons) [reply src/FirewallManager.cs:40]: From Landaire's rust code:
```
        CoInitializeEx(None, COINIT_MULTITHREADED)
            .ok()
            .context("CoInitializeEx")?;
```

As for the result, [MSDN states](https://learn.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-coinitializeex#return-value) that the following for `CoInitializeEx()`:

| Return Code | Description |
| --- | --- |
| S_OK | The COM library was initialized successfully on this thread. |
| S_FALSE | The COM library is already initialized on this thread. |

I assumed both of these scenarios are nominal for our use case.  Happy to be wrong.

Comment (tuxuser) [reply src/FirewallManager.cs:40]: Thanks for the wakeup call, you are absolutely right!

**Review by tuxuser — APPROVED** [review]

**Pull request merged**